resolver/src/p2p/fetcher.rs

```diff
@@ -94,133 +148,205 @@ impl<E: Clock + GClock + Rng + Metrics, P: PublicKey, Key: Span, NetS: Sender<Pu
+
+            // Try each peer in order until one succeeds
+            for peer in peers {
+                // Generate request ID
+                let id = self.next_id();
+
+                // Record request before sending
+                let now = self.context.current();
+                let deadline = now.checked_add(self.timeout).expect("time overflowed");
+
+                // Try to send
+                let result = sender
+                    .send(
+                        Recipients::One(peer.clone()),
+                        wire::Message {
+                            id,
+                            payload: wire::Payload::Request(key.clone()),
+                        },
+                        self.priority_requests,
+                    )
+                    .await;
+
+                match result {
+                    Ok(sent) if !sent.is_empty() => {
+                        // Success - remove from pending and add to active
+                        self.requests_sent.inc(Status::Success);
+                        self.pending.remove(&key);
+                        self.active.put(id, deadline);
+                        self.requests.insert(
+                            id,
+                            ActiveRequest {
+                                key: key.clone(),
+                                peer,
+                                start: now,
+                            },
+                        );
+                        self.key_to_id.insert(key, id);
+                        return;
+                    }
+                    Ok(_) => {
+                        // Empty result - peer was rate-limited by Sender, try next
+                        self.requests_sent.inc(Status::Dropped);
+                        debug!(?peer, "send returned empty (rate limited)");
+                        continue;
+                    }
+                    Err(err) => {
+                        // Send error - update performance and try next peer
+                        self.requests_sent.inc(Status::Failure);
+                        debug!(?err, ?peer, "send failed");
+                        self.update_performance(&peer, self.timeout);
+                        continue;
                     }
-                    // If a request with targets fails to find a peer, other keys may still be fetchable
-                }
-                Err(Error::NoEligibleParticipants) => {
-                    // If a request with no valid targets exists (blocked or targets no longer allowed), we
-                    // should skip it (may eventually become fetchable if the peer set changes).
                 }
             }
-        }
 
-        // Send request if we found a key to fetch
-        if let Some((key, peer, id)) = selected {
-            self.pending.remove(&key);
-            return self.send_request(sender, key, peer, id).await;
+            // All peers exhausted for this key, try next key
```

Comment:
The purpose of the previous loop was to stop early if all peers were rate-limited (determined by an untargeted key not being sent to anyone). It would be great to retain this (so we don't need to loop over all keys for each request if no one is able).